From 638b339e7fc9c40a4c06dd13c4e084c32b98b627 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Wed, 18 Jun 2025 10:14:42 +0200
Subject: [PATCH 194/985] net: dovetail: do stricter checks in
 skb_oob_dma_addr()

Make sure the storage area carried by the skb does not belong to an
oob page pool before attempting to fetch a DMA address from its page
descriptor. In the same move, drop the check on the recycling flag
which is orthogonal to this issue.

NOTE: we still don't check whether skb->head is actually paged memory.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 include/linux/skbuff.h | 19 +++++++++++++++++--
 1 file changed, 17 insertions(+), 2 deletions(-)

diff --git a/include/linux/skbuff.h b/include/linux/skbuff.h
index c4f041dd2..245e35782 100644
--- a/include/linux/skbuff.h
+++ b/include/linux/skbuff.h
@@ -5257,6 +5257,7 @@ ssize_t skb_splice_from_iter(struct sk_buff *skb, struct iov_iter *iter,
 #ifdef CONFIG_NET_OOB
 
 #include <net/page_pool/helpers.h>
+#include <linux/skbuff_ref.h>
 
 extern unsigned int sysctl_max_oob_skb;
 
@@ -5318,12 +5319,26 @@ static inline struct skb_shared_oob *skb_shinfo_oob(struct sk_buff *skb)
 	return &skb_shinfo(skb)->oob_shinfo;
 }
 
+/**
+ * skb_oob_dma_addr - Return the DMA address of a pre-mapped buffer
+ * obtained from an oob pool.
+ *
+ * DMA_MAPPING_ERROR is returned if the skb is dataless, or its
+ * storage area does not belong to an oob page pool.
+ */
 static inline dma_addr_t skb_oob_dma_addr(const struct sk_buff *skb)
 {
-	if (!skb->head || !skb_is_oob_managed(skb) || !skb->pp_recycle)
+	struct page *page = skb->head ? virt_to_page(skb->head) : NULL;
+	struct page_pool *pool;
+
+	if (!page)
+		return DMA_MAPPING_ERROR;
+
+	pool = napi_pp_get_pool(page_to_netmem(page));
+	if (!pool || !page_pool_is_oob(pool))
 		return DMA_MAPPING_ERROR;
 
-	return page_pool_get_dma_addr(virt_to_page(skb->head));
+	return page_pool_get_dma_addr(page);
 }
 
 bool recycle_skb_oob(struct sk_buff *skb);
-- 
2.30.2

